### PR TITLE
Util--Signature-ecsign-ecrecover-functionality

### PR DESCRIPTION
**Pull request by ScottyPoi**
Re: #1961 

Util:

- Adds `recovery` to `ECDSASignature` interface.
- Creates a new method `calculateSigRecoveryFromV(v: bigint)`
  - This includes a calculation for recovery based on the `recovery = chainId * 2 + 35` formula.
- Refactors `fromRpcSig()` to return `ECDSASignature` with `recovery`.
- Updates relevant  `signature.spec.ts` tests for new interface, methods.

TX:
For legacy transactions, `Transaction._validateTxV()` was updated to filter for hardforks that include EIP-150

Client:
- Small changes were made regarding `miner` and `pendingBlock` in response to failing tests.

### CODE REVIEW

File: packages/util/src/signature.ts [hunk 1]
```diff
@@ -39,6 +39,21 @@ function calculateSigRecovery(v: bigint, chainId?: bigint): bigint {
...
+  if (v === BigInt(0) || v === BigInt(1)) {
+    return v
+  } else {
+    const _v = Number(v)
+    if (Number.isInteger((0 - 35 - _v) / -2)) {
+      return BigInt(0)
+    } else if (Number.isInteger((1 - 35 - _v) / -2)) {
+      return BigInt(1)
+    } else {
+      throw new Error('Invalid v value')
```
Comment (holgerd77) [inline packages/util/src/signature.ts:52]: Can we here remain in the BigInt context? `v` *can* theoretically (actually: this also already happened practically) get in the > MAX_SAFE_INTEGER range, when there is a very large chain ID (we once had this for a testnet), so arithmetics here should take place in BigInt.

Just tested, there is the modulo operator `%` also working in `BigInt`, so: `BigInt(5) % BigInt(2)` gave me `1n`, so it should be possible with something like this to get this working in plain `BigInt` as well.

File: packages/util/src/signature.ts [hunk 2]
```diff
@@ -39,6 +39,21 @@ function calculateSigRecovery(v: bigint, chainId?: bigint): bigint {
   return v - (chainId * BigInt(2) + BigInt(35))
 }
 
+export function calculateSigRecoveryFromV(v: bigint) {
```
Comment (holgerd77) [inline packages/util/src/signature.ts:42]: Totally a fan of having this exported in this new form, seems pretty useful to me, but can we then (also generally makes sense) some explicit test cases for this method itself, atm all tests are implicit (another method using this function is tested)?

So some edge cases like 1 (I guess we at some point decided/agreed that 0 is not a practical/used chain ID which we can ignore), a mid range even and odd value, and (related to the other comment) a very high v exceeding MAX_SAFE_INTEGER.

**Review by holgerd77 — CHANGES_REQUESTED** [review]
Hi Scotty,
this looks totally great, thanks for tackling and love the cleanliness in this sensitve parts of our code base.

Two comments two address, also there are some `vm-api` CI tests still failing directly related to the new `recovery` property in the signature. This would need fixes as well.

Side note: branch update / rebase also necessary.

File: packages/util/src/signature.ts [hunk 3]
```diff
@@ -8,6 +8,7 @@ export interface ECDSASignature {
   v: bigint
   r: Buffer
   s: Buffer
+  recovery: bigint
 }
```
Comment (holgerd77) [inline packages/util/src/signature.ts:12]: Another short review round 🙂: can you please add some code docs to this interface?

File: packages/util/src/signature.ts (hunk 2 above)
Comment (holgerd77) [inline packages/util/src/signature.ts:42]: Also here, it would be great if you can add a TSDocs description.

File: packages/util/src/signature.ts [hunk 4]
```diff
@@ -39,6 +39,21 @@ function calculateSigRecovery(v: bigint, chainId?: bigint): bigint {
   return v - (chainId * BigInt(2) + BigInt(35))
 }
```
Comment (holgerd77) [inline packages/util/src/signature.ts:40]: I guess we can (and also: should) delete this method since this is redundant now.

File: packages/util/src/signature.ts [hunk 5]
```diff
@@ -128,6 +143,7 @@ export const fromRpcSig = function (sig: string): ECDSASignature {
     throw new Error('Invalid signature length')
   }
 
+  const recovery: bigint = calculateSigRecoveryFromV(v)
```
Comment (holgerd77) [inline packages/util/src/signature.ts:146]: Small nit: you do not need to explicity type this, this is inferred by the method call.

**Review by holgerd77 — CHANGES_REQUESTED** [review]
Short second round. 🙂

File: packages/util/src/signature.ts (hunk 4 above)
Comment (ScottyPoi) [reply packages/util/src/signature.ts]: By deleting this method, I had to shift `ecrecover` to use the new `calculateSigRecoveryFromV()` instead of the deleted `calculateSigRecovery()`

`calculateSigRecovery` had a clause that if the `chainId` parameter was undefined, the method would throw.  Without chainId as a parameter at all, `calculateSigRecoveryFromV()` will always return.

Several tests in util and tx relied on this method throwing an error.

**Comment by holgerd77** [comment]
This is getting too short-termed, so we'll rather add in Beta 2, no problem though. 🙂

**Comment by holgerd77** [comment]
Giving this a "Blocked" label right now, can be removed again after the Beta 1 releases.

File: packages/client/lib/miner/miner.ts [hunk 6]
```diff
@@ -257,7 +257,7 @@ export class Miner {
       },
       blockOpts: {
         cliqueSigner,
-        hardforkByBlockNumber: true,
```
[the remark this answers is not in the document]
Comment (ScottyPoi) [reply packages/client/lib/miner/miner.ts:260]: Do you know why it was on?

This caused a test to fail that involved building a `Fake Chain`, and arbitrarily assigning hardforks to block numbers.
```
    const customChainParams = {
      hardforks: [
        { name: 'chainstart', block: 0 },
        { name: 'berlin', block: 2 },
        { name: 'london', block: 3 },
      ],
    }
```


https://github.com/ethereumjs/ethereumjs-monorepo/blob/26e1890d800851bd9907a1978ff83e92594840a1/packages/client/test/miner/miner.spec.ts#L373

with `hardforkByBlockNumber` on, it would always revert back to `chainstart`, even with the   `common` option set to  `this.config.chainCommon`

[the remark this answers is not in the document]
Comment (ScottyPoi) [reply packages/client/lib/miner/miner.ts:260]: I think Ryan designed the test.

If there's a strong reason for having `hardforkByBlockNumber` on, I could instead look into changing `VM.buildBlock()` to ignore it if the `common` option is set?

[the remark this answers is not in the document]
Comment (g11tech) [reply packages/client/lib/miner/miner.ts:260]: its on i think to make sure the correct block is produced according to hardfork at that current block number (could be a side chain/non-canonical chain and loading hardfork at that block number makes sense)

but how should your PR which doesn't look relevant to this, affect this logic, thats the first thing to understand why this spec test is failing.

[the remark this answers is not in the document]
Comment (g11tech) [reply packages/client/lib/miner/miner.ts:260]: most likely the mockchain might not be build properly because of some validation failing, that the miner is building block from `0` where it seems like it should build post `3`. I will try this PR out tomorrow locally to confirm this hypo.

[the remark this answers is not in the document]
Comment (ScottyPoi) [reply packages/client/lib/miner/miner.ts:260]: Thanks I spent more time with it today and didn't get any further

[the remark this answers is not in the document]
Comment (g11tech) [reply packages/client/lib/miner/miner.ts:260]: i think I have resolved the spec test, so turning this back on :+1:

File: packages/util/src/signature.ts [hunk 7]
```diff
@@ -43,10 +43,9 @@ export function calculateSigRecoveryFromV(v: bigint) {
   if (v === BigInt(0) || v === BigInt(1)) {
     return v
   } else {
-    const _v = Number(v)
-    if (Number.isInteger((0 - 35 - _v) / -2)) {
+    if ((0n - 35n - v) % 2n === 0n) {
       return BigInt(0)
-    } else if (Number.isInteger((1 - 35 - _v) / -2)) {
+    } else if ((1n - 35n - v) % 2n === 0n) {
```
Comment (holgerd77) [inline packages/util/src/signature.ts]: Please do not use `n` notation for BigInt for now, we decided at some point to keep with the `BigInt()` constructor for tool compatibility reasons.

File: packages/tx/src/legacyTransaction.ts [hunk 8]
```diff
@@ -296,13 +296,7 @@ export default class Transaction extends BaseTransaction<Transaction> {
     this._validateHighS()
 
     try {
-      return ecrecover(
-        msgHash,
-        v!,
-        bigIntToUnpaddedBuffer(r!),
-        bigIntToUnpaddedBuffer(s!),
-        this.supports(Capability.EIP155ReplayProtection) ? this.common.chainId() : undefined
```
Comment (holgerd77) [inline packages/tx/src/legacyTransaction.ts:304]: Please do not do any changes to the transaction library in this PR, this is out of the scope of the task and by far too dangerous.

File: packages/util/src/signature.ts (hunk 4 above)
Comment (holgerd77) [reply packages/util/src/signature.ts]: If this causes any problems than in  doubt *keep* the method. This was just a quick shot from me assuming this internal method is not needed any more.

**Review by holgerd77 — CHANGES_REQUESTED** [review]

**Comment by holgerd77** [comment]
Replaced by #2012, will close.
